File: behavioral/visitor/code_analyzer/python/code_analyzer/syntax_check_visitor.py
from .visitor import Visitor
from .ast_nodes import (
    FunctionDefinitionNode,
    VariableDeclarationNode,
    AssignmentStatementNode,
    IfStatementNode,
    ExpressionStatementNode,
    ExpressionNode, # Needed for assignment RHS check (though simplified here)
)
from typing import List, Set

class SyntaxCheckVisitor(Visitor):
    """Performs simple syntax checks, like tracking declared variables."""
    def __init__(self):
        self._errors: List[str] = []
        # Stack of sets to manage variable scopes
        self._scope_stack: List[Set[str]] = [set()] # Start with global scope

    def get_errors(self) -> List[str]:
        return self._errors

    def _add_error(self, message: str):
        self._errors.append(message)

    def _push_scope(self):
        self._scope_stack.append(set())

    def _pop_scope(self):
        if len(self._scope_stack) > 1: # Avoid popping the global scope
            self._scope_stack.pop()

    def _declare_variable(self, name: str, node_info="variable"):
        current_scope = self._scope_stack[-1]
        if name in current_scope:
            self._add_error(f"Syntax Error: Identifier '{name}' already declared as a {node_info} in this scope.")
        else:
            current_scope.add(name)

    def _is_declared(self, name: str) -> bool:
        # Check from current scope outwards
        for scope in reversed(self._scope_stack):
            if name in scope:
                return True
        return False

    # --- Visitor Methods ---

    def visit_function_definition(self, element: FunctionDefinitionNode):
        # Scope entry/exit should be handled externally by the caller
        # around the accept() call for the FunctionDefinitionNode.
        # This method now focuses only on traversing the body.
        for stmt in element.body:
             stmt.accept(self)
        # No scope pop here - done externally by exit_function_scope

    def visit_variable_declaration(self, element: VariableDeclarationNode):
        # Declare in the current scope
        self._declare_variable(element.name)
        # Check initializer if present (e.g., ensure variables in expression are declared)
        if element.initializer:
             # Simplified: We don't deeply check expression variables here
             pass
        # No need to push/pop scope here

    def visit_assignment_statement(self, element: AssignmentStatementNode):
        # Check if the target variable exists
        if not self._is_declared(element.target_variable):
            self._add_error(f"Syntax Error: Variable '{element.target_variable}' used before declaration.")
        # Check RHS expression variables if needed (simplified)
        pass

    def visit_if_statement(self, element: IfStatementNode):
        # Check condition variables (simplified)
        # Visit the 'then' branch within its own scope
        self.enter_block_scope()
        for stmt in element.then_branch:
            stmt.accept(self) # Visit children within the 'then' scope
        self.exit_block_scope()

        if element.else_branch:
            # Visit the 'else' branch within its own scope
            self.enter_block_scope()
            for stmt in element.else_branch:
                stmt.accept(self) # Visit children within the 'else' scope
            self.exit_block_scope()

    def visit_expression_statement(self, element: ExpressionStatementNode):
        # Check variables/function calls within the expression (simplified)
        pass

    # --- Scope Management Helpers ---
    def enter_function_scope(self, function_name: str, parameters: list[str]):
        # Declare function name in the current (likely global) scope
        self._declare_variable(function_name, "function")
        # Push a new scope for the function body
        self._push_scope()
        # Declare parameters in the new function scope
        for param in parameters:
            self._declare_variable(param, "parameter")

    def exit_function_scope(self):
        self._pop_scope()

    def enter_block_scope(self):
        self._push_scope()

    def exit_block_scope(self):
        self._pop_scope()

    def reset(self):
        """Resets the visitor's state (errors and scope stack)."""
        self._errors = []
        self._scope_stack = [set()] # Initialize with global scope

    def get_errors(self):
        return self._errors

File: behavioral/visitor/code_analyzer/typescript/code_analyzer/syntax_check_visitor.ts
import { Visitor } from "./visitor";
import {
  FunctionDefinitionNode,
  VariableDeclarationNode,
  AssignmentStatementNode,
  IfStatementNode,
  ExpressionStatementNode,
  // ExpressionNode // Needed if checking expression internals
} from "./ast_nodes";

// Performs simple syntax checks, like tracking declared variables using scopes.
export class SyntaxCheckVisitor implements Visitor {
  private errors: string[] = [];
  // Stack of Sets to manage variable scopes (using Set for efficient lookup)
  private scopeStack: Set<string>[] = [new Set<string>()]; // Start with global scope

  getErrors(): string[] {
    return this.errors;
  }

  private addError(message: string): void {
    this.errors.push(message);
  }

  private pushScope(): void {
    this.scopeStack.push(new Set<string>());
  }

  private popScope(): void {
    if (this.scopeStack.length > 1) {
      // Avoid popping the global scope
      this.scopeStack.pop();
    }
  }

  private declareVariable(name: string, nodeInfo: string = "variable"): void {
    const currentScope = this.scopeStack[this.scopeStack.length - 1];
    if (currentScope?.has(name)) {
      this.addError(
        `Syntax Error: Identifier '${name}' already declared as a ${nodeInfo} in this scope.`
      );
    } else {
      currentScope?.add(name);
    }
  }

  private isDeclared(name: string): boolean {
    // Check from current scope outwards
    for (let i = this.scopeStack.length - 1; i >= 0; i--) {
      if (this.scopeStack[i]?.has(name)) {
        return true;
      }
    }
    return false;
  }

  // --- Visitor Methods ---

  visitFunctionDefinition(element: FunctionDefinitionNode): void {
    // Declare function name in the current (outer) scope
    this.declareVariable(element.name, "function");
    // Push a new scope for the function body
    this.pushScope();
    // Declare parameters in the new scope
    element.parameters.forEach((param) =>
      this.declareVariable(param, "parameter")
    );

    // Explicitly traverse the function body within the new scope
    element.body.forEach((stmt) => stmt.accept(this));

    // Pop the scope after visiting the body
    this.popScope();
  }

  visitVariableDeclaration(element: VariableDeclarationNode): void {
    // Declare in the current scope
    this.declareVariable(element.name);
    // Check initializer if present
    element.initializer?.accept(this); // Visit initializer
  }

  visitAssignmentStatement(element: AssignmentStatementNode): void {
    // Check if the target variable exists
    if (!this.isDeclared(element.targetVariable)) {
      this.addError(
        `Syntax Error: Variable '${element.targetVariable}' used before declaration.`
      );
    }
    // Check RHS expression variables if needed
    element.value.accept(this); // Visit RHS value
  }

  visitIfStatement(element: IfStatementNode): void {
    // Check condition variables
    element.condition.accept(this); // Visit condition

    // Push scope for the 'then' branch
    this.pushScope();
    // Explicitly traverse the 'then' branch statements
    element.thenBranch.forEach((stmt) => stmt.accept(this));
    // Pop scope after 'then' branch traversal
    this.popScope();

    if (element.elseBranch) {
      // Push scope for the 'else' branch
      this.pushScope();
      // Explicitly traverse the 'else' branch statements
      element.elseBranch.forEach((stmt) => stmt.accept(this));
      // Pop scope after 'else' branch traversal
      this.popScope();
    }
  }

  visitExpressionStatement(element: ExpressionStatementNode): void {
    // Check variables/function calls within the expression
    element.expression.accept(this); // Visit expression
  }

  // Method to clear state for reuse
  public reset(): void {
    this.errors = [];
    this.scopeStack = [new Set<string>()]; // Re-initialize with global scope (using Set as originally intended)
  }
}
